### PR TITLE
Implementing the change proposed in #502

### PR DESCRIPTION
**Pull request by goerch**

### CODE REVIEW

File: src/types.jl [hunk 1]
```diff
@@ -250,12 +250,12 @@ function Frame(framecode::FrameCode, framedata::FrameData, pc=1, caller=nothing)
...
 """
-    frame = Frame(mod::Module, src::CodeInfo)
+    frame = Frame(mod::Module, src::CodeInfo; generator::Bool, optimize::Bool)
 
 Construct a `Frame` to evaluate `src` in module `mod`.
 """
-function Frame(mod::Module, src::CodeInfo)
-    framecode = FrameCode(mod, src)
+function Frame(mod::Module, src::CodeInfo; generator=false, optimize=true)
+    framecode = FrameCode(mod, src, generator=generator, optimize=optimize)
```
Comment (KristofferC) [inline src/types.jl:258]:
```suggestion
    framecode = FrameCode(mod, src; generator, optimize)
```

maybe? :)

File: src/types.jl [hunk 2]
```diff
@@ -250,12 +250,12 @@ function Frame(framecode::FrameCode, framedata::FrameData, pc=1, caller=nothing)
...
 """
-    frame = Frame(mod::Module, src::CodeInfo)
+    frame = Frame(mod::Module, src::CodeInfo; generator::Bool, optimize::Bool)
 
 Construct a `Frame` to evaluate `src` in module `mod`.
 """
-function Frame(mod::Module, src::CodeInfo)
-    framecode = FrameCode(mod, src)
+function Frame(mod::Module, src::CodeInfo; generator=false, optimize=true)
+    framecode = FrameCode(mod, src; generator, optimize)
```
Comment (aviatesk) [inline src/types.jl:258]:
```suggestion
function Frame(mod::Module, src::CodeInfo; kwargs...)
    framecode = FrameCode(mod, src; kwargs...)
```
would look nicer?

File: src/types.jl [hunk 3]
```diff
@@ -250,12 +250,12 @@ function Frame(framecode::FrameCode, framedata::FrameData, pc=1, caller=nothing)
     end
 end
 """
-    frame = Frame(mod::Module, src::CodeInfo)
+    frame = Frame(mod::Module, src::CodeInfo; generator::Bool, optimize::Bool)
```
Comment (aviatesk) [inline src/types.jl:253]:
```suggestion
    frame = Frame(mod::Module, src::CodeInfo; kwargs...)
```

**Comment by aviatesk** [comment]
We want to add testcase that demonstrates how these added option can be used. I just come up with something like this:
```julia
@testset "interpretation of unoptimized frame" begin
    let # should be able to interprete nested calls within `:foreigncall` expressions
        # even if `JuliaInterpreter.optimize!` doesn't flatten them
        M = Module()
        lwr = Meta.@lower M begin
            global foo = @ccall strlen("foo"::Cstring)::Csize_t
            foo == 3
        end
        src = lwr.args[1]::Core.CodeInfo
        frame = Frame(M, src; optimize=false)
        @test length(frame.framecode.src.code) == length(src.code)
        @test JuliaInterpreter.finish_and_return!(frame, true)

        M = Module()
        lwr = Meta.@lower M begin
            strp = Ref{Ptr{Cchar}}(0)
            fmt = "hi+%hhd-%hhd-%hhd-%hhd-%hhd-%hhd-%hhd-%hhd-%hhd-%hhd-%hhd-%hhd-%hhd-%hhd-%hhd-%.1f-%.1f-%.1f-%.1f-%.1f-%.1f-%.1f-%.1f-%.1f\n"
            len = @ccall asprintf(
                strp::Ptr{Ptr{Cchar}},
                fmt::Cstring,
                ; # begin varargs
                0x1::UInt8, 0x2::UInt8, 0x3::UInt8, 0x4::UInt8, 0x5::UInt8, 0x6::UInt8, 0x7::UInt8, 0x8::UInt8, 0x9::UInt8, 0xa::UInt8, 0xb::UInt8, 0xc::UInt8, 0xd::UInt8, 0xe::UInt8, 0xf::UInt8,
                1.1::Cfloat, 2.2::Cfloat, 3.3::Cfloat, 4.4::Cfloat, 5.5::Cfloat, 6.6::Cfloat, 7.7::Cfloat, 8.8::Cfloat, 9.9::Cfloat,
            )::Cint
            str = unsafe_string(strp[], len)
            @ccall free(strp[]::Cstring)::Cvoid
            str == "hi+1-2-3-4-5-6-7-8-9-10-11-12-13-14-15-1.1-2.2-3.3-4.4-5.5-6.6-7.7-8.8-9.9\n"
        end
        src = lwr.args[1]::Core.CodeInfo
        frame = Frame(M, src; optimize=false)
        @test length(frame.framecode.src.code) == length(src.code)
        @test JuliaInterpreter.finish_and_return!(frame, true)
    end
end
```

**Comment by goerch** [comment]
> finish_and_return! currently errors, therefore I commented it for now.

Here are the errors:

```
WARNING: Method definition foobar() in module Main at C:\Users\Win10\Documents\GitHub\JuliaInterpreter.jl\test\interpret.jl:779 overwritten at C:\Users\Win10\Documents\GitHub\JuliaInterpreter.jl\test\interpret.jl:780.
interpretation of unoptimized frame: Error During Test at C:\Users\Win10\Documents\GitHub\JuliaInterpreter.jl\test\interpret.jl:864
  Test threw exception
  Expression: JuliaInterpreter.finish_and_return!(frame, true)
  invalid lookup expr Core.svec(Cstring)
  Stacktrace:
    [1] error(::String, ::Expr)
      @ Base .\error.jl:42
    [2] lookup_expr(frame::Frame, e::Expr)
      @ JuliaInterpreter C:\Users\Win10\Documents\GitHub\JuliaInterpreter.jl\src\interpret.jl:25
    [3] collect_args(frame::Frame, call_expr::Expr; isfc::Bool)
      @ JuliaInterpreter C:\Users\Win10\Documents\GitHub\JuliaInterpreter.jl\src\interpret.jl:60
    [4] evaluate_foreigncall(frame::Frame, call_expr::Expr)
      @ JuliaInterpreter C:\Users\Win10\Documents\GitHub\JuliaInterpreter.jl\src\interpret.jl:151
    [5] eval_rhs(recurse::Any, frame::Frame, node::Expr)
      @ JuliaInterpreter C:\Users\Win10\Documents\GitHub\JuliaInterpreter.jl\src\interpret.jl:420
    [6] step_expr!(recurse::Any, frame::Frame, node::Any, istoplevel::Bool)
      @ JuliaInterpreter C:\Users\Win10\Documents\GitHub\JuliaInterpreter.jl\src\interpret.jl:572
    [7] step_expr!
      @ C:\Users\Win10\Documents\GitHub\JuliaInterpreter.jl\src\interpret.jl:627 [inlined]
    [8] finish!(recurse::Any, frame::Frame, istoplevel::Bool)
      @ JuliaInterpreter C:\Users\Win10\Documents\GitHub\JuliaInterpreter.jl\src\commands.jl:14
    [9] finish_and_return!
      @ C:\Users\Win10\Documents\GitHub\JuliaInterpreter.jl\src\commands.jl:30 [inlined]
   [10] finish_and_return!(frame::Frame, istoplevel::Bool)
      @ JuliaInterpreter C:\Users\Win10\Documents\GitHub\JuliaInterpreter.jl\src\commands.jl:34
   [11] macro expansion
      @ C:\Users\Win10\AppData\Local\Programs\Julia-1.7.0\share\julia\stdlib\v1.7\Test\src\Test.jl:445 [inlined]
   [12] macro expansion
      @ C:\Users\Win10\Documents\GitHub\JuliaInterpreter.jl\test\interpret.jl:864 [inlined]
   [13] macro expansion
      @ C:\Users\Win10\AppData\Local\Programs\Julia-1.7.0\share\julia\stdlib\v1.7\Test\src\Test.jl:1283 [inlined]
   [14] top-level scope
      @ C:\Users\Win10\Documents\GitHub\JuliaInterpreter.jl\test\interpret.jl:854
interpretation of unoptimized frame: Error During Test at C:\Users\Win10\Documents\GitHub\JuliaInterpreter.jl\test\interpret.jl:884
  Test threw exception
  Expression: JuliaInterpreter.finish_and_return!(frame, true)
  invalid lookup expr Core.svec(Core.apply_type(Ptr, Core.apply_type(Ptr, Cchar)), Cstring, UInt8, UInt8, UInt8, UInt8, UInt8, UInt8, UInt8, UInt8, UInt8, UInt8, UInt8, 
UInt8, UInt8, UInt8, UInt8, Cfloat, Cfloat, Cfloat, Cfloat, Cfloat, Cfloat, Cfloat, Cfloat, Cfloat)
  Stacktrace:
    [1] error(::String, ::Expr)
      @ Base .\error.jl:42
    [2] lookup_expr(frame::Frame, e::Expr)
      @ JuliaInterpreter C:\Users\Win10\Documents\GitHub\JuliaInterpreter.jl\src\interpret.jl:25
    [3] collect_args(frame::Frame, call_expr::Expr; isfc::Bool)
      @ JuliaInterpreter C:\Users\Win10\Documents\GitHub\JuliaInterpreter.jl\src\interpret.jl:60
    [4] evaluate_foreigncall(frame::Frame, call_expr::Expr)
      @ JuliaInterpreter C:\Users\Win10\Documents\GitHub\JuliaInterpreter.jl\src\interpret.jl:151
    [5] eval_rhs(recurse::Any, frame::Frame, node::Expr)
      @ JuliaInterpreter C:\Users\Win10\Documents\GitHub\JuliaInterpreter.jl\src\interpret.jl:420
    [6] step_expr!(recurse::Any, frame::Frame, node::Any, istoplevel::Bool)
      @ JuliaInterpreter C:\Users\Win10\Documents\GitHub\JuliaInterpreter.jl\src\interpret.jl:572
    [7] step_expr!
      @ C:\Users\Win10\Documents\GitHub\JuliaInterpreter.jl\src\interpret.jl:627 [inlined]
    [8] finish!(recurse::Any, frame::Frame, istoplevel::Bool)
      @ JuliaInterpreter C:\Users\Win10\Documents\GitHub\JuliaInterpreter.jl\src\commands.jl:14
    [9] finish_and_return!
      @ C:\Users\Win10\Documents\GitHub\JuliaInterpreter.jl\src\commands.jl:30 [inlined]
   [10] finish_and_return!(frame::Frame, istoplevel::Bool)
      @ JuliaInterpreter C:\Users\Win10\Documents\GitHub\JuliaInterpreter.jl\src\commands.jl:34
   [11] macro expansion
      @ C:\Users\Win10\AppData\Local\Programs\Julia-1.7.0\share\julia\stdlib\v1.7\Test\src\Test.jl:445 [inlined]
   [12] macro expansion
      @ C:\Users\Win10\Documents\GitHub\JuliaInterpreter.jl\test\interpret.jl:884 [inlined]
   [13] macro expansion
      @ C:\Users\Win10\AppData\Local\Programs\Julia-1.7.0\share\julia\stdlib\v1.7\Test\src\Test.jl:1283 [inlined]
   [14] top-level scope
      @ C:\Users\Win10\Documents\GitHub\JuliaInterpreter.jl\test\interpret.jl:854
```

**Comment by goerch** [comment]
Interesting question: shouldn't `collect_args` get an additional recurse argument to use for `lookup_or_eval`?

```
function collect_args(frame::Frame, call_expr::Expr; isfc::Bool=false)
    args = frame.framedata.callargs
    resize!(args, length(call_expr.args))
    mod = moduleof(frame)
    args[1] = isfc ? resolvefc(frame, call_expr.args[1]) : @lookup(mod, frame, call_expr.args[1])
    for i = 2:length(args)
        if isexpr(call_expr.args[i], :call)
            args[i] = lookup_or_eval(finish_and_return!, frame, call_expr.args[i])
        else
            args[i] = @lookup(mod, frame, call_expr.args[i])
        end
    end
    return args
end
```

File: src/interpret.jl [hunk 4]
```diff
@@ -130,25 +132,29 @@ function resolvefc(frame, @nospecialize(expr))
...
-function collect_args(frame::Frame, call_expr::Expr; isfc::Bool=false)
+function collect_args(@nospecialize(recurse), frame::Frame, call_expr::Expr; isfc::Bool=false)
     args = frame.framedata.callargs
     resize!(args, length(call_expr.args))
     mod = moduleof(frame)
     args[1] = isfc ? resolvefc(frame, call_expr.args[1]) : @lookup(mod, frame, call_expr.args[1])
     for i = 2:length(args)
-        args[i] = @lookup(mod, frame, call_expr.args[i])
+        if isexpr(call_expr.args[i], :call)
+            args[i] = lookup_or_eval(recurse, frame, call_expr.args[i])
```
Comment (timholy) [inline src/interpret.jl:142]: With this change, we might need to test
```julia
iscallexpr(ex::Expr) = ex.head === :call
@interpret iscallexpr(:(sin(3.14))
```
We need to be sure we distinguish a call expression intended as an argument from one produced by lowering.

**Review by timholy — COMMENTED** [review]
If the ambiguity about call expressions can be resolved, this LGTM.

**Comment by timholy** [comment]
Thanks @goerch!

**Pull request merged**